src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp

```diff
@@ -0,0 +1,162 @@
+  * @param maxNumTokens The maximum number of tokens in the strings of the 
+                        input dataset.
+  * @param dictionarySize The size of the dictionary (not used).
+  */
+  template<typename OutputType>
+  static void InitMatrix(std::vector<std::vector<OutputType> >& output,
+                         size_t datasetSize,
+                         size_t /*maxNumTokens*/,
+                         size_t dictionarySize)
+  {
+    output.resize(datasetSize, std::vector<OutputType> (dictionarySize,0));
+  }
+
+  /** 
+  * The function performs the TfIdf encoding algorithm i.e. it writes
+  * the encoded token to the ouput.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param value The encoded token.
+  * @param row The row number at which the encoding is performed.
+  * @param col The row token number at which the encoding is performed.
+  */
+  template<typename MatType>
+  void Encode(MatType& output, size_t value, size_t row, size_t /*col*/)
+  {
+    // Important since Mapping starts from 1 whereas allowed column value is 0.
+    output(row, value-1) = (tokenCount[row][value - 1] / row_size[row]) * 
+        std::log10(output.n_rows / idfdict[value-1]);
+  }
+
+  /** 
+  * The function performs the TfIdf encoding algorithm i.e. it writes
+  * the encoded token to the ouput.
+  * Overload function to accepted vector<vector<size_t>> as output type.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param value The encoded token.
+  * @param row The row number at which the encoding is performed.
+  * @param col The row token number at which the encoding is performed.
+  */
+  template<typename OutputType>
+  void Encode(std::vector<std::vector<OutputType> >& output, size_t value,
+                     size_t row, size_t /*col*/)
+  {
+    // Important since Mapping starts from 1 whereas allowed column value is 0.
+    output[row][value-1] = (tokenCount[row][value - 1] / row_size[row]) * 
+        std::log10(output.size() / idfdict[value-1]);
+  }
+
+  /**
+   * Serialize the class to the given archive.
+   */
+  template<typename Archive>
+  void serialize(Archive& /* ar */, const unsigned int /* version */)
+  {
+    // Nothing to serialize.
+  }
+
+  /*
+  * The function is used to create the datastrcutre will be important to find
+  * out idfvalue of words, and then wrtiting the output based on their count.
+  *
+  * @param row The row number at which the encoding is performed.
+  * @param numToken The count of token parsed till now.
+  * @param value The encoded token.
+  */
+   void PreprocessToken(size_t row, size_t /*numTokens*/,
+                       size_t value)
+  {
+    if(row>=tokenCount.size())
+    {
+      row_size.push_back(0);
+      tokenCount.push_back(std::unordered_map<size_t, double>());
+    }
+    tokenCount.back()[value-1]++;
+    if(tokenCount.back()[value-1]==1)
+      idfdict[value-1]++;
```

Comment:
Looks like you can move `idfdict[value-1]++;` to the previous `if` body.

Reply:
That will run only once for a row, and hence i am not sure, that will give us correct output,Did i miss something ?

Reply:
The previous condition holds true once for each row too.

Reply:
> That will run only once for a row, and hence i am not sure, that will give us correct output,Did i miss something ?

The above thing was for the previous if

`idfdict[value-1]++;`: I want it to run once every row for every unique token, and hence i need to compare it on the basis of token, And that is why i am checking whether the value is 1, since it should run only once every row

Reply:
I see. You're right.